File: video_looper.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstdlib> // For system()
#include <cstdio>  // For remove()
#include <filesystem> // For path operations (C++17)
#include <sstream> // For string stream

// Function to check if a file exists
bool file_exists(const std::filesystem::path& p) {
    return std::filesystem::exists(p) && std::filesystem::is_regular_file(p);
}

// Helper function to escape paths for shell commands if needed (basic version)
std::string quote_path(const std::filesystem::path& p) {
    // On Windows, paths with spaces often need double quotes.
    // On Linux/macOS, single quotes can be more robust, or careful escaping.
    // For simplicity, using double quotes. More robust escaping might be needed for complex paths.
    return "\"" + p.string() + "\"";
}

bool create_looped_video_with_audio(
    const std::filesystem::path& input_video_path,
    const std::filesystem::path& input_audio_path,
    int num_loops,
    const std::filesystem::path& output_video_path) {

    if (!file_exists(input_video_path)) {
        std::cerr << "Error: Input video file not found at '" << input_video_path << "'" << std::endl;
        return false;
    }
    if (!file_exists(input_audio_path)) {
        std::cerr << "Error: Input audio file not found at '" << input_audio_path << "'" << std::endl;
        return false;
    }
    if (num_loops <= 0) {
        std::cerr << "Error: Number of loops must be a positive integer." << std::endl;
        return false;
    }

    std::filesystem::path temp_concat_file_path = "ffmpeg_concat_list.txt";
    std::filesystem::path temp_looped_video_path = "temp_looped_video_only.mp4";

    // 1. Create the concatenation file for FFmpeg
    std::cout << "Creating concatenation file..." << std::endl;
    { // Scope for ofstream to ensure it's closed before FFmpeg uses the file
        std::ofstream concat_file(temp_concat_file_path);
        if (!concat_file.is_open()) {
            std::cerr << "Error: Could not create temporary concat file: " << temp_concat_file_path << std::endl;
            return false;
        }
        std::filesystem::path absolute_input_video_path = std::filesystem::absolute(input_video_path);
        for (int i = 0; i < num_loops; ++i) {
            // FFmpeg concat demuxer needs 'file' directive. Paths should be escaped if they contain special characters.
            // Using absolute paths is generally safer for the concat demuxer with -safe 0.
            // The paths written to the file list itself should NOT be quoted if they don't contain problematic characters
            // for the list parser, but the input_video_path itself might need cleaning if it has single quotes.
            // For simplicity, assuming paths are well-behaved or FFmpeg's -safe 0 handles them with abspath.
            concat_file << "file '" << absolute_input_video_path.string() << "'\n";
        }
        concat_file.close();
    }
    std::cout << "Concatenation file created: " << temp_concat_file_path << std::endl;

    // 2. Loop the video using FFmpeg concat demuxer
    std::cout << "\nLooping video..." << std::endl;
    std::ostringstream ffmpeg_loop_command_ss;
    ffmpeg_loop_command_ss << "ffmpeg -y -f concat -safe 0 -i " << quote_path(temp_concat_file_path)
                           << " -an -c:v copy "
                           << quote_path(temp_looped_video_path);
    std::string ffmpeg_loop_command = ffmpeg_loop_command_ss.str();

    std::cout << "Executing: " << ffmpeg_loop_command << std::endl;
    int loop_ret_code = system(ffmpeg_loop_command.c_str());
    if (loop_ret_code != 0) {
        std::cerr << "Error: FFmpeg video looping failed with code " << loop_ret_code << std::endl;
        std::remove(temp_concat_file_path.c_str()); // Clean up
        return false;
    }
    std::cout << "Video looping successful." << std::endl;

    // 3. Combine looped video with audio
    std::cout << "\nCombining video and audio..." << std::endl;
    std::ostringstream ffmpeg_combine_command_ss;
    ffmpeg_combine_command_ss << "ffmpeg -y -i " << quote_path(temp_looped_video_path)
                              << " -i " << quote_path(std::filesystem::absolute(input_audio_path)) // Use absolute for audio too
                              << " -map 0:v:0 -map 1:a:0 -c:v copy -c:a aac -b:a 192k -shortest "
                              << quote_path(output_video_path);
    std::string ffmpeg_combine_command = ffmpeg_combine_command_ss.str();

    std::cout << "Executing: " << ffmpeg_combine_command << std::endl;
    int combine_ret_code = system(ffmpeg_combine_command.c_str());
    if (combine_ret_code != 0) {
        std::cerr << "Error: FFmpeg video/audio combining failed with code " << combine_ret_code << std::endl;
        std::remove(temp_concat_file_path.c_str());
        std::remove(temp_looped_video_path.c_str()); // Clean up
        return false;
    }
    std::cout << "Video and audio combination successful." << std::endl;

    // 4. Clean up temporary files
    std::cout << "\nCleaning up temporary files..." << std::endl;
    if (std::remove(temp_concat_file_path.c_str()) != 0) {
        std::cerr << "Warning: Could not remove temporary concat file: " << temp_concat_file_path << std::endl;
    }
    if (std::remove(temp_looped_video_path.c_str()) != 0) {
        std::cerr << "Warning: Could not remove temporary looped video file: " << temp_looped_video_path << std::endl;
    }
    std::cout << "Cleanup finished." << std::endl;

    std::cout << "\nSuccessfully created '" << output_video_path << "' with " << num_loops << " loops and selected audio." << std::endl;
    return true;
}

int main() {
    // --- Configuration ---
    std::filesystem::path input_video = "input.mp4";
    std::filesystem::path input_audio = "audio.mp3";
    int number_of_loops = 3;
    std::filesystem::path output_video = "final_looped_video_cpp.mp4";
    // --- End Configuration ---

    std::cout << "Starting video processing application..." << std::endl;
    std::cout << "Input Video: " << std::filesystem::absolute(input_video) << std::endl;
    std::cout << "Input Audio: " << std::filesystem::absolute(input_audio) << std::endl;
    std::cout << "Number of Loops: " << number_of_loops << std::endl;
    std::cout << "Output Video: " << std::filesystem::absolute(output_video) << std::endl;

    // Check if input files exist before calling the main function
    // This is also done inside the function, but good for an early exit.
    if (!file_exists(input_video)) {
        std::cerr << "Error: Main - Input video file '" << input_video
                  << "' not found. Please create it or update the path." << std::endl;
        std::cerr << "To create a dummy 1-second black video (input.mp4):" << std::endl;
        std::cerr << "ffmpeg -y -f lavfi -i color=c=black:s=1280x720:d=1 -f lavfi -i anullsrc=channel_layout=stereo:sample_rate=44100 -c:v libx264 -c:a aac -t 1 input.mp4" << std::endl;
        return 1;
    }
    if (!file_exists(input_audio)) {
        std::cerr << "Error: Main - Input audio file '" << input_audio
                  << "' not found. Please create it or update the path." << std::endl;
        std::cerr << "To create a dummy 5-second silent audio (audio.mp3):" << std::endl;
        std::cerr << "ffmpeg -y -f lavfi -i anullsrc=channel_layout=stereo:sample_rate=44100 -t 5 audio.mp3" << std::endl;
        return 1;
    }


    bool success = create_looped_video_with_audio(input_video, input_audio, number_of_loops, output_video);

    if (success) {
        std::cout << "\nVideo processing finished successfully." << std::endl;
    } else {
        std::cout << "\nVideo processing failed." << std::endl;
        return 1;
    }

    return 0;
}

File: VideoLooper.js
const fs = require('fs').promises; // For async file operations
const path = require('path');
const { exec } = require('child_process'); // Can use spawn for better stream handling if needed

// Helper to run shell commands and return a Promise
function executeCommand(command) {
    return new Promise((resolve, reject) => {
        console.log(`Executing: ${command}`);
        const process = exec(command, (error, stdout, stderr) => {
            if (stdout) console.log('Stdout:\n', stdout);
            if (stderr) console.error('Stderr:\n', stderr);
            if (error) {
                console.error(`Execution error for command "${command}":`, error);
                reject(error);
                return;
            }
            resolve({ stdout, stderr });
        });
    });
}

async function createLoopedVideoWithAudio(inputVideoPath, inputAudioPath, numLoops, outputVideoPath) {
    const absInputVideoPath = path.resolve(inputVideoPath);
    const absInputAudioPath = path.resolve(inputAudioPath);
    const absOutputVideoPath = path.resolve(outputVideoPath);

    try {
        await fs.access(absInputVideoPath); // Check if input video exists
    } catch (e) {
        console.error(`Error: Input video file not found at '${absInputVideoPath}'`);
        return false;
    }
    try {
        await fs.access(absInputAudioPath); // Check if input audio exists
    } catch (e) {
        console.error(`Error: Input audio file not found at '${absInputAudioPath}'`);
        return false;
    }

    if (numLoops <= 0) {
        console.error("Error: Number of loops must be a positive integer.");
        return false;
    }

    const tempConcatFilePath = path.resolve("ffmpeg_concat_list.txt");
    const tempLoopedVideoPath = path.resolve("temp_looped_video_only.mp4");

    try {
        // 1. Create the concatenation file
        console.log(`Creating concatenation file: ${tempConcatFilePath}`);
        let concatFileContent = "";
        for (let i = 0; i < numLoops; ++i) {
            // FFmpeg concat demuxer needs 'file' directive.
            // For paths with spaces, they should be quoted *within the file list entry*.
            // The path.resolve should give an absolute path, which helps.
            // Replacing backslashes for Windows paths used in ffmpeg file lists.
            const sanitizedPath = absInputVideoPath.replace(/\\/g, '/');
            concatFileContent += `file '${sanitizedPath}'\n`;
        }
        await fs.writeFile(tempConcatFilePath, concatFileContent);
        console.log("Concatenation file created.");

        // 2. Loop the video (no audio)
        console.log("\nLooping video...");
        // Quote paths for the shell command itself
        const loopCommand = `ffmpeg -y -f concat -safe 0 -i "${tempConcatFilePath}" -an -c:v copy "${tempLoopedVideoPath}"`;
        await executeCommand(loopCommand);
        console.log("Video looping successful.");

        // 3. Combine looped video with audio
        console.log("\nCombining video and audio...");
        const combineCommand = `ffmpeg -y -i "${tempLoopedVideoPath}" -i "${absInputAudioPath}" -map 0:v:0 -map 1:a:0 -c:v copy -c:a aac -b:a 192k -shortest "${absOutputVideoPath}"`;
        await executeCommand(combineCommand);
        console.log("Video and audio combination successful.");

        console.log(`\nSuccessfully created '${absOutputVideoPath}' with ${numLoops} loops and selected audio.`);
        return true;

    } catch (error) {
        console.error("An error occurred during video processing:", error.message || error);
        if(error.stderr) console.error("FFMPEG Stderr:", error.stderr);
        if(error.stdout) console.log("FFMPEG Stdout:", error.stdout);
        return false;
    } finally {
        // 4. Clean up temporary files
        console.log("\nCleaning up temporary files...");
        try {
            await fs.unlink(tempConcatFilePath);
            console.log(`Deleted: ${tempConcatFilePath}`);
        } catch (e) {
            console.warn(`Warning: Could not remove temporary concat file: ${tempConcatFilePath} (${e.message})`);
        }
        try {
            await fs.unlink(tempLoopedVideoPath);
            console.log(`Deleted: ${tempLoopedVideoPath}`);
        } catch (e) {
            console.warn(`Warning: Could not remove temporary looped video file: ${tempLoopedVideoPath} (${e.message})`);
        }
        console.log("Cleanup finished.");
    }
}

// --- Main execution ---
(async () => {
    // --- Configuration ---
    const inputVideo = "input.mp4";
    const inputAudio = "audio.mp3";
    const numberOfLoops = 225;
    const outputVideo = "final_looped_video_js.mp4";
    // --- End Configuration ---

    console.log("Starting video processing application (Node.js)...");
    console.log(`Input Video: ${path.resolve(inputVideo)}`);
    console.log(`Input Audio: ${path.resolve(inputAudio)}`);
    console.log(`Number of Loops: ${numberOfLoops}`);
    console.log(`Output Video: ${path.resolve(outputVideo)}`);

    try {
        await fs.access(inputVideo);
    } catch (e) {
        console.error(`Error: Main - Input video file '${inputVideo}' not found. Please create it or update the path.`);
        console.error("To create a dummy 1-second black video (input.mp4):");
        console.error("ffmpeg -y -f lavfi -i color=c=black:s=1280x720:d=1 -f lavfi -i anullsrc=channel_layout=stereo:sample_rate=44100 -c:v libx264 -c:a aac -t 1 input.mp4");
        return;
    }
     try {
        await fs.access(inputAudio);
    } catch (e) {
        console.error(`Error: Main - Input audio file '${inputAudio}' not found. Please create it or update the path.`);
        console.error("To create a dummy 5-second silent audio (audio.mp3):");
        console.error("ffmpeg -y -f lavfi -i anullsrc=channel_layout=stereo:sample_rate=44100 -t 5 audio.mp3");
        return;
    }


    const success = await createLoopedVideoWithAudio(inputVideo, inputAudio, numberOfLoops, outputVideo);

    if (success) {
        console.log("\nVideo processing finished successfully.");
    } else {
        console.log("\nVideo processing failed.");
    }
})();